pip/pip-338.md

```diff
@@ -0,0 +1,172 @@
+               if (redirectUrl == null) {
+                   log.error("Redirected cluster's service url is not configured");
+                   throw new RestException(Response.Status.PRECONDITION_FAILED,
+                           "Redirected cluster's service url is not configured.");
+               }
+               String lookupPath = topicName.isV2() ? LOOKUP_PATH_V2 : LOOKUP_PATH_V1;
+               String path = String.format("%s%s%s?authoritative=%s",
+                       redirectUrl, lookupPath, topicName.getLookupName(), newAuthoritative);
+               path = listenerName == null ? path : path + "&listenerName=" + listenerName;
+               redirect = new URI(path);
+           }
+    
+```
+
+## Need to have lookupListener
+The Pulsar broker has the concept of advertised listeners representing broker endpoints that are discoverable by, and accessible to, Pulsar clients. In addition, one listener may be designated as the internal listener, to be used for broker-to-broker communication. Listeners may be understood as named routes to the broker. If a listener name is not specified, the response contains the endpoint for the internal listener, or the first listener for the protocol.
+
+But the challenge arises when it considers the same internalListener for lookup requests redirects. This may result in returning an unresolvable broker service address to the clients.
+It may not be possible to pass the listenerName from the clients, which consequently may lead to cluster downtime.
+
+it might not be feasible to have the listerner config at the client side in every tech stack or connector.
+Also,  Currently, there is no option to pass the listener while using the admin APIs. As admin APIs can be called from an external network as well, the use of the internal listener’s broker service URL can lead all admin operations to get affected.
+
+Moreover, as per current code provided above, even if the client provides a listenerName/listener header, the redirect urls are taken from the internal listener or the first listener with http protocol which may not be a client-resolvable http address.
+
+# Proposed Solution
+To fix inconsistency with all protocol’s service url, we need to include all broker addresses associated with the listener name while returning the lookupResult, rather than solely the service URL
+
+Also, Introduce a broker-level property to determine the appropriate listener for lookup calls if listenerName is missing from the client request. This can help deal with the side effects of not having listerName on the client side when multiple advertisedListeners are being used.
+
+Also, this can help to determine the broker service URL that needs to be returned in case of lookup redirects and also for using Pulsar admin API calls from outside the network.
+
+This will help in having a more transparent listener identifier for internal and external communications.
```

Comment:
Thinking out aloud here, I haven't come to a conclusion yet.

In [PIP-95](https://github.com/apache/pulsar/issues/12040) there's a solution that I'd prefer instead:
> The broker shall be configurable to open a server socket on numerous bind addresses, and to associate each socket with a listener name. When a lookup request arrives on a particular socket, the broker shall use the associated listener by default. Note that an explicit listener name in the request parameters would take precedence.

The main problem I have is that there is already a design for solving the problem, but that has never been implemented in Pulsar. 

I acknowledge that the benefit of the `lookupListenerName` solution would be that it could be more practical to implement. In that case, the `internalListenerName` should be made effective so that it is used in all broker-to-broker connections within the cluster. 

The limitation of the `lookupListenerName` solution is that it assumes that there's only one external listener for the cluster. The PIP-95 solution of "use a unique bind address for each listener." would be better since there wouldn't be any such limits that could cause problems for more advanced use cases.

Reply:
I agree that there are inconsistencies in design and implementation. Maybe there were missing things in design of PIP-91 and 95 itself as well.

I do not think a bind address mapping is going to solve this. For instance, the same http scheme/protocol based external facing url leads to multiple different protocol changes in output - 

http based lookup call can either result in the client using the pulsar protocol as the output, or an http based redirect url as an output.

http based admin calls always end up with either 200 or http based redirect url.

Both the above cases may have the same entry point in terms of bind address.

Please correct me if my understanding is wrong here in the examples.

Reply:
To add to this, I actually would refer to what kafka has been doing for a long while.. our proposal is actually very close to kafka [KIP 103 ](https://cwiki.apache.org/confluence/display/KAFKA/KIP-103%3A+Separation+of+Internal+and+External+traffic) which PIP-95 was actually supposed to be doing..

Reply:
@grssam I hope the comment https://github.com/apache/pulsar/pull/22039#discussion_r1487282934 explains how this can be solved for the Pulsar Admin API http redirects.

Reply:
What about http lookup leading to pulsar protocol being used? The common use case for clients is to use the http load balancer address for initial service discovery, which means that the lookup service used by the client will be the HttpLookupService ..  Now in this case, when the actual lookup happens, if its a redirect, the http redirect url is used but in case its a 200 response, the pulsar protocol address is used. Thus - leading to different output listener for the same bind address.

Reply:
> Now in this case, when the actual lookup happens, if its a redirect, the http redirect url is used but in case its a 200 response, the pulsar protocol address is used.

The code at https://github.com/apache/pulsar/blob/3158fd3550f9e3a0b2c0316c92265318b209f4f5/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java#L189-L197 , the call to `findRedirectLookupResultAsync` method ignores that listener name that is passed in in the LookupOptions parameter. That looks like a bug and not an implementation gap. :)

Reply:
Those are exactly the inconsistencies that we are trying to fix here.

Along with a default lookup listener so that we are not "forced" to just send a custom listener name from the client side without which the system does not work at all. A default shouldn't be something that is this broken for users that without overriding, things don't work at all.